modules/axhal/src/platform/pc_x86/acpi.rs

```diff
@@ -0,0 +1,333 @@
+    }
+
+    fn read_pci_u16(&self, segment: u16, bus: u8, device: u8, function: u8, offset: u16) -> u16 {
+        let paddr = unsafe {
+            ACPI.get_pci_config_regions_addr(segment, bus, device, function)
+                .unwrap()
+        };
+        let vaddr = phys_to_virt(PhysAddr::from(paddr as usize)).as_ptr() as *const u16;
+        let address = unsafe { vaddr.add(offset as usize) };
+        unsafe { address.read_volatile() }
+    }
+
+    fn read_pci_u32(&self, segment: u16, bus: u8, device: u8, function: u8, offset: u16) -> u32 {
+        let paddr = unsafe {
+            ACPI.get_pci_config_regions_addr(segment, bus, device, function)
+                .unwrap()
+        };
+        let vaddr = phys_to_virt(PhysAddr::from(paddr as usize)).as_ptr() as *const u32;
+        let address = unsafe { vaddr.add(offset as usize) };
+        unsafe { address.read_volatile() }
+    }
+
+    fn write_pci_u8(
+        &self,
+        segment: u16,
+        bus: u8,
+        device: u8,
+        function: u8,
+        offset: u16,
+        value: u8,
+    ) {
+        let paddr = unsafe {
+            ACPI.get_pci_config_regions_addr(segment, bus, device, function)
+                .unwrap()
+        };
+        let vaddr = phys_to_virt(PhysAddr::from(paddr as usize)).as_mut_ptr();
+        let address = unsafe { vaddr.add(offset as usize) };
+        unsafe { address.write_volatile(value) }
+    }
+
+    fn write_pci_u16(
+        &self,
+        segment: u16,
+        bus: u8,
+        device: u8,
+        function: u8,
+        offset: u16,
+        value: u16,
+    ) {
+        let paddr = unsafe {
+            ACPI.get_pci_config_regions_addr(segment, bus, device, function)
+                .unwrap()
+        };
+        let vaddr = phys_to_virt(PhysAddr::from(paddr as usize)).as_mut_ptr() as *mut u16;
+        let address = unsafe { vaddr.add(offset as usize) };
+        unsafe { address.write_volatile(value) }
+    }
+
+    fn write_pci_u32(
+        &self,
+        segment: u16,
+        bus: u8,
+        device: u8,
+        function: u8,
+        offset: u16,
+        value: u32,
+    ) {
+        let paddr = unsafe {
+            ACPI.get_pci_config_regions_addr(segment, bus, device, function)
+                .unwrap()
+        };
+        let vaddr = phys_to_virt(PhysAddr::from(paddr as usize)).as_mut_ptr() as *mut u32;
+        let address = unsafe { vaddr.add(offset as usize) };
+        unsafe { address.write_volatile(value) }
+    }
+}
+
+struct Acpi {
+    rsdp: AcpiTables<LocalAcpiHandler>,
+    aml_context: AmlContext,
+}
+
+/// irq model used in ACPI
+#[allow(dead_code)]
+enum X86IrqModel {
+    /// PIC model
+    PIC,
+    /// APIC model
+    APIC,
+}
+
+impl Acpi {
+    pub unsafe fn new() -> Self {
+        Acpi {
+            rsdp: AcpiTables::search_for_rsdp_bios(LocalAcpiHandler).unwrap(),
+            aml_context: AmlContext::new(Box::new(LocalAmlHandler), DebugVerbosity::None),
```

Comment:
Sorry, we can't for now. Because both `acpi` and `aml` crate rely on `alloc`. Crate `acpi` is used to parse ACPI table and Crate `aml` is used to invoke ACPI method. Both are need to finish the job of getting irq routing from PCI device pin to APIC pin.

Although the developers of `acpi` and `aml` crate offer another crate `rsdp` for environment without heap alloctors, it only finds and parses `RSDP` and cannot parse the data and methods in ACPI table.

Crate `aml` is hard to eliminate since we need to invoke ACPI method `_PRT`  to get irq routing from PCI device pin to APIC pin.

----------------

However, if we accept a legacy way, we may get irq number of PCI device without ACPI completely.

It seems that PCI device irq pin connects to two APIC pins in many motherboards (e.g. in current arceos qemu environment, virtio-net PCI device pinA is connected to APIC pin both 11 and 22) , one (e.g. 11) is for compatibility with PIC, and can  be readed from PCI configuration space under `IRQ Line` field.

Since arceos disables PIC and uses APIC only, I think it is better to use the pin specialized for APIC (which needs to get from ACPI) and neglect the pin for PIC.

By the way, using the PIC one(e.g. 11) has another (small) programming problem. 
Current `virtio_drivers` crate on github, which is responsible  for parsing virtio PCI configuration space, doesn't have a method to read `IRQ Line` field. We may need to add this method if we choose the legacy solution.

Any suggestions are welcome and appreciated.😁

Reply:
Can we drop the legacy IRQ support and use MSI instead?

Reply:
Yes, we can. It seems to be a much better solution. I will try it later.

> Can we drop the legacy IRQ support and use MSI instead?